controller/engine_image_controller.go

```diff
@@ -394,72 +394,67 @@ func (ic *EngineImageController) syncNodeDeploymentMap(engineImage *longhorn.Eng
 
 // handleAutoUpgradeEngineImageToDefaultEngineImage automatically upgrades volume's engine image to default engine image when it is applicable
 func (ic *EngineImageController) handleAutoUpgradeEngineImageToDefaultEngineImage() error {
-	concurrentAutomaticEngineUpgradePerNodeLimit, err := ic.ds.GetSettingAsInt(types.SettingNameConcurrentAutomaticEngineUpgradePerNodeLimit)
+	defaultEngineImage, err := ic.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
 	if err != nil {
 		return err
 	}
-	if concurrentAutomaticEngineUpgradePerNodeLimit <= 0 {
+	defaultEngineImageResource, err := ic.ds.GetEngineImage(types.GetEngineImageChecksumName(defaultEngineImage))
+	if err != nil {
+		return err
+	}
+	// Make sure that only the owner of the default engine image does the upgrade
+	if defaultEngineImageResource.Status.OwnerID != ic.controllerID {
 		return nil
 	}
```

Comment:
Why not directly check if the current engineimage is the default one? I prefer not to consider the ownership issue here.

Reply:
I agree with this, move the check outside into the main sync loop.
Then rename this function to `handleAutoUpgradeEngineImage(newImage *longhorn.EngineImage, limit int)` pass the new image (default) and limit to this function. Then the only function that needs to retrieve settings is the main sync loop in the case where it's the owner of the default image.

https://github.com/longhorn/longhorn-manager/blob/0a2885634f0f7fcf159721132d53b501804248f3/controller/engine_image_controller.go#L356

Reply:
Thanks @shuo-wu 
Fixed!

Reply:
Sounds reasonable if it's best effort, then keep it in there. 
Reasons for exposing the parameters are the ability to do unit tests against that function, but it's not important for the moment so keep it the way you had it with the settings and everything included in that single function.